core/src/main/java/jenkins/scm/RunWithSCM.java

```diff
@@ -0,0 +1,130 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.scm;
+
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.User;
+import hudson.scm.ChangeLogSet;
+import hudson.util.AdaptedIterator;
+import jenkins.util.SystemProperties;
+import org.kohsuke.stapler.export.Exported;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.AbstractSet;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * @since FIXME
+ */
+public interface RunWithSCM<JobT extends Job<JobT, RunT> & Queue.Task,
+        RunT extends Run<JobT, RunT> & RunWithSCM<JobT,RunT> & Queue.Executable> {
+
+    List<ChangeLogSet<? extends ChangeLogSet.Entry>> getChangeSets();
+
+    @CheckForNull
+    Set<String> getCulpritIds();
+
+    boolean shouldCalculateCulprits();
+
+    /**
+     * List of users who committed a change since the last non-broken build till now.
+     *
+     * <p>
+     * This list at least always include people who made changes in this build, but
+     * if the previous build was a failure it also includes the culprit list from there.
+     *
+     * @return
+     *      can be empty but never null.
+     */
+    @Exported
+    @Nonnull default Set<User> getCulprits() {
+        if (shouldCalculateCulprits()) {
+            return calculateCulprits();
+        }
+
+        return new AbstractSet<User>() {
+            public Iterator<User> iterator() {
+                return new AdaptedIterator<String,User>(getCulpritIds().iterator()) {
+                    protected User adapt(String id) {
+                        return User.get(id);
+                    }
+                };
+            }
+
+            public int size() {
+                return getCulpritIds().size();
+            }
+        };
+    }
+
+    @SuppressWarnings("unchecked")
+    default Set<User> calculateCulprits() {
+        Set<User> r = new HashSet<User>();
+        RunT p = ((RunT)this).getPreviousCompletedBuild();
+        if (p != null && ((RunT)this).isBuilding()) {
+            Result pr = p.getResult();
+            if (pr != null && pr.isWorseThan(Result.SUCCESS)) {
+                // we are still building, so this is just the current latest information,
+                // but we seems to be failing so far, so inherit culprits from the previous build.
+                // isBuilding() check is to avoid recursion when loading data from old Hudson, which doesn't record
+                // this information
+                r.addAll(p.getCulprits());
+            }
+        }
+        for (ChangeLogSet<? extends ChangeLogSet.Entry> c : getChangeSets()) {
+            for (ChangeLogSet.Entry e : c)
```

Comment:
So long as we are breaking `git blame` and friends anyway, may as well reformat cleanly.